src/libraries/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -136,49 +137,47 @@ public BitArray(bool[] values)
             ref byte value = ref Unsafe.As<bool, byte>(ref MemoryMarshal.GetArrayDataReference<bool>(values));
             if (Vector512.IsHardwareAccelerated)
             {
-                for (; i <= (uint)values.Length - Vector512<byte>.Count; i += (uint)Vector512<byte>.Count)
+                for (; i <= values.Length - Vector512<byte>.Count; i += Vector512<byte>.Count)
                 {
-                    Vector512<byte> vector = Vector512.LoadUnsafe(ref value, i);
+                    Vector512<byte> vector = Vector512.LoadUnsafe(ref value, (uint)i);
                     Vector512<byte> isFalse = Vector512.Equals(vector, Vector512<byte>.Zero);
 
-                    ulong result = isFalse.ExtractMostSignificantBits();
-                    m_array[i / 32u] = (int)(~result & 0x00000000FFFFFFFF);
-                    m_array[(i / 32u) + 1] = (int)((~result >> 32) & 0x00000000FFFFFFFF);
+                    ulong result = ~(isFalse.ExtractMostSignificantBits());
```

Comment:
This is less efficient, you should rather have this be `ulong result = (~Vector512.IsZero(vector)).ExtractMostSignificantBits()`

Reply:
It's almost always better to do the negations and similar on the comparison or vector directly, as it allows it to become a direct `cmpneq` instead of staying a `cmpeq` followed by `not` or similar

Reply:
1. There is no `pcmpneq`, only `pcmpeq`.
2. AVX512 does have `vptestmb` or `vpcmpub`, but neither `(~Vector512.IsZero(value)).ExtractMostSignificantBits()` nor `~(Vector512.IsZero(value).ExtractMostSignificantBits())` are optimised to use them (latest daily build).

`(~Vector512.IsZero(value)).ExtractMostSignificantBits()`:
```asm
vmovups  zmm0, zmmword ptr [rcx]
vptestnmb k1, zmm0, zmm0
vpmovm2b zmm0, k1
vpternlogd zmm0, zmm0, zmm0, 85
vpmovb2m k1, zmm0
kmovq    rax, k1
vzeroupper
ret
```

`~(Vector512.IsZero(value).ExtractMostSignificantBits())`:
```asm
vmovups  zmm0, zmmword ptr [rcx]
vptestnmb k1, zmm0, zmm0
kmovq    rax, k1
not      rax
vzeroupper
ret
```

Currently, `~(Vector512.IsZero(value).ExtractMostSignificantBits())` looks better.